wrf-cmake.rb

```diff
@@ -0,0 +1,73 @@
+# WRF-CMake (https://github.com/WRF-CMake/WRF).
+# Copyright 2019 M. Riechert and D. Meyer. Licensed under the MIT License.
+
+class WrfCmake < Formula
+  desc "The Weather Research and Forecasting (WRF) model with CMake support"
+  homepage "https://github.com/WRF-CMake/WRF"
+  url "https://github.com/WRF-CMake/WRF.git",
+      :branch => "wrf-cmake",
+      :commit => "8b9edc944f0cbbfb22966850ae3724f617f0f705"
+  version "4.1.wrf-cmake"
+  sha256 "37814ee7bfe7077cba8bd0175258ef763380fce3e7d8aab2ea8360a902a11043"
+  head "https://github.com/WRF-CMake/WRF.git", :branch => "wrf-cmake"
+
+  depends_on "cmake" => :build
+  depends_on "gcc" # for gfortran
+  depends_on "jasper"
+  depends_on "libpng"
+  depends_on "netcdf"
+  depends_on "open-mpi"
+
+  resource "WPS" do
+    url "https://github.com/WRF-CMake/WPS.git",
+        :branch => "wps-cmake",
+        :commit => "d918dc6da17b177aa1fe0082866274968f45a171"
+  end
+
+  # Note: If you get an internal compiler error when building WRF,
+  # then you're likely running out of memory. To reduce memory usage,
+  # set HOMEBREW_MAKE_JOBS=1.
+
+  def install
+    (buildpath/"WPS").install resource("WPS")
+
+    # Prevent CMake from finding & using ifort
+    ENV["FC"] = Formula["gcc"].opt_bin/"gfortran"
+
+    wrf_args = std_cmake_args.map do |s|
+      s.sub /-DCMAKE_INSTALL_PREFIX.*$/, "-DCMAKE_INSTALL_PREFIX=#{prefix}/wrf"
+    end
+    wrf_args << "-DNESTING=basic" << "-DMODE=dmpar"
+
+    mkdir "build" do
+      system "cmake", "..", *wrf_args
+      system "make", "install"
+    end
+
+    # Ideally, WPS should be a separate formula, but it relies on the build tree of WRF,
+    # which is not installed. Therefore, we build it here and install it in the WPS subfolder
+    # of WRF.
+    wps_args = std_cmake_args + %W[
+      -DCMAKE_INSTALL_PREFIX=#{prefix}/wps
+      -DWRF_DIR=#{buildpath}/build
```

Comment:
I'm curious: Why is this necessary? What needs to be found from the build directory? Are you linking in a library target or something?

Reply:
Yes, static libraries. WPS depends on WRF. We don't want to support the use case of really installing WRF as a first-class library, since this would be something new.

Reply:
I don't really get why WPS is a separate package then. It feels like the worst of both worlds. But I understand people are slow to adopt new things, and you may need to gently coax them into the 21st century... So if you think this is necessary (for now at least) to increase the odds of upstream accepting this improvement, then I'm all for it.

An idea for splitting the homebrew formula would be to have the wrf formula copy the needed libraries into a directory in the install tree (like `libs_for_wps`) then point the WPS install to that directory.

IMO, splitting out WPS can probably be tabled due to the unusual installation tree and strange (but less strange than upstream, 😄) build practices. Hopefully in the future options in CMake can be added to start:

* using normal binary names
* installing to canonical locations
* installing libraries needed by dependent packages
* install a cmake config file

Have you been in touch with upstream throughout this process? It's possible that they may either a) be completely un-receptive to change or b) be more willing than you realize to modernize things. It's hard to read minds. But again, I defer to you and your judgement, you're much more familiar with the community than I am.

Reply:
I fully agree with your thinking. Currently we are conservative to have it as a compatible drop-in replacement. In the future this may change, and we will probably make suggestions to upstream about that, but given that WRF is decades old, it takes time. We've been in touch with upstream but interest was still low at that time. We'll re-engage once the paper is out and we have a bit more momentum.
I think all your changes are good to go now. Thanks again for helping us out here and being so patient! The world needs more people like you :)

Reply:
It would be pretty straight forwards to have a compatibility mode and canonical mode that could be toggled in CMake. They could ease into it with the default being compatibility, then eventually they could switch it.

Thanks for the kind words. Fortran is a great language and it's good to see efforts like this one to modernize, make more robust, and canonicalize great, but admittedly old skool, scientific software like WRF.